CONTRIBUTING.md

```diff
@@ -0,0 +1,101 @@
+branch that starts at `next` and is specific to the pull request. Once you have
+submitted a pull request, people will start reviewing it and providing feedback.
+If you don't hear anything for a while feel free to bring it to the attention of
+people in [IRC](https://freenetproject.org/irc.html) or the [mailing
+list](https://emu.freenetproject.org/cgi-bin/mailman/listinfo/devl) - we may
+have missed it.
+
+Code review helps improve code quality, ensures that multiple people know the
+codebase, serves as a defense against introducing malicious code, and makes it
+infeasable to pressure people into contributing malicious code. Its goal is
+producing software that is readable, correct, and sufficiently efficient.
+
+Breaking API creates an immense amount of work for developers of other projects,
+so it is *not* something to be done lightly. If you feel you must make a change
+that breaks API and cannot maintain backwards compatibility, please first raise
+the issue with the community - the [mailing list](https://emu.freenetproject.org/cgi-bin/mailman/listinfo/devl)
+and [IRC](https://freenetproject.org/irc.html) are good places to contact us.
+
+# Standards
+
+Before submitting a pull request, please make sure that modified lines meet the
+project [coding standards](https://google-styleguide.googlecode.com/svn/trunk/javaguide.html)
+and that the commit messages meet the standards:
+
+## Commit messages
+
+From the git commit man page:
+
+> begin the commit message with a single short (less than 50 character) line
+> summarizing the change, followed by a blank line and then a more thorough
+> description. The text up to the first blank line in a commit message is
+> treated as the commit title...
+
+Except for proper nouns, only the first character of the title may be
+capitalized. It may start with the class / module the change applies to
+followed by a colon if it helps to explain the change and give it context. The
+main part of it is a short imperative description - a command given to the
+codebase. The title is at most 72 characters and does not end with a period. The
+second line is blank. If the message contains additional prose description on
+the third and subsequent lines, it is wrapped to 72 characters. Long lines that
+do not split well, such as URLs or stack traces, are an exception to this.
+
+    [Context:] command given to the codebase
+
+For example:
+
+    Update default bookmark editions
+    
+    Additional description of what the change does, why it is a good idea,
+    and alternate solutions decided against, if any, goes here. For some
+    changes this might not be necessary.
+
+For more discussion see the [git patch submission documentation](https://git.kernel.org/cgit/git/git.git/tree/Documentation/SubmittingPatches#n87).
+
+Text editors can be configured to assist in formatting messages this way, and
+git packages sometimes ship with such configuration.
+
+## Commit separation
```

Comment:
In addition to or instead of the current content?

On Thu, Apr 9, 2015, 9:11 AM Arne Babenhauserheide <notifications@github.com>
wrote:

> In CONTRIBUTING.md
> <https://github.com/freenet/fred/pull/338#discussion_r28059548>:
>
> > +    [Context:] command given to the codebase
> > +
> > +For example:
> > +
> > +    Update default bookmark editions
> > +
> > +    Additional description of what the change does, why it is a good idea,
> > +    and alternate solutions decided against, if any, goes here. For some
> > +    changes this might not be necessary.
> > +
> > +For more discussion see the [git patch submission documentation](https://git.kernel.org/cgit/git/git.git/tree/Documentation/SubmittingPatches#n87).
> > +
> > +Text editors can be configured to assist in formatting messages this way, and
> > +git packages sometimes ship with such configuration.
> > +
> > +## Commit separation
>
> How about saying “we avoid destroying history except if that is necessary
> to avoid making review a bottleneck”?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/freenet/fred/pull/338/files#r28059548>.
>